tests/d3d12_device.c

```diff
@@ -356,13 +559,75 @@ void test_format_support(void)
     ok(!format_support.Support2 || format_support.Support2 == D3D12_FORMAT_SUPPORT2_TILED,
             "Got unexpected support2 %#x.\n", format_support.Support2);
 
-    for (i = 0; i < ARRAY_SIZE(depth_stencil_formats); ++i)
+    memset(&feature_levels, 0, sizeof(feature_levels));
+    feature_levels.NumFeatureLevels = ARRAY_SIZE(all_feature_levels);
+    feature_levels.pFeatureLevelsRequested = all_feature_levels;
+    hr = ID3D12Device_CheckFeatureSupport(device, D3D12_FEATURE_FEATURE_LEVELS, 
+        &feature_levels, sizeof(feature_levels));
+    ok(hr == S_OK, "Got unexpected hr %#x.\n", hr);
+
+    for (k = 0; k < ARRAY_SIZE(dxgi_format_list); ++k)
+    {
+        if (dxgi_format_list[k].feature_level == feature_levels.MaxSupportedFeatureLevel)
+            break;
+    }
+
+    for (i = 0; i < ARRAY_SIZE(all_formats); ++i)
     {
+        required_but_fails = false;
+        unspecified_format = false;
+        optional_format = false;
+
+        for (j = 0; j < ARRAY_SIZE(known_required_but_fails); ++j)
+        {
+            if (known_required_but_fails[j] == all_formats[i])
+            {
+                required_but_fails = true;
+                break;
+            }
+        }
+
+        if (!required_but_fails)
+        {
+            /* Check if the format is unspecified or optional */
+            for (j = 0; j < ARRAY_SIZE(dxgi_format_list[k].unspecified_formats); ++j)
+            {
+                // Fixed size list with only part of the list filled
+                if (dxgi_format_list[k].unspecified_formats[j] == DXGI_FORMAT_UNKNOWN)
+                    break;
+                if (dxgi_format_list[k].unspecified_formats[j] == all_formats[i])
+                {
+                    unspecified_format = true;
+                    break;
+                }
+            }
+
+            if (!unspecified_format)
+            {
+                for (j = 0; j < ARRAY_SIZE(dxgi_format_list[k].optional_formats); ++j)
+                {
+                    // Fixed size list with only part of the list filled
+                    if (dxgi_format_list[k].optional_formats[j] == DXGI_FORMAT_UNKNOWN)
+                        break;
+                    if (dxgi_format_list[k].optional_formats[j] == all_formats[i])
+                    {
+                        optional_format = true;
+                        break;
+                    }
+                }
+            }
+        }
+
         memset(&format_support, 0, sizeof(format_support));
-        format_support.Format = depth_stencil_formats[i];
+        format_support.Format = all_formats[i];
         hr = ID3D12Device_CheckFeatureSupport(device, D3D12_FEATURE_FORMAT_SUPPORT,
                 &format_support, sizeof(format_support));
-        ok(hr == S_OK, "Got unexpected hr %#x.\n", hr);
+        if (unspecified_format)
```

Comment:
Shouldn't unspecified_format return E_FAIL?

Reply:
D3D12 runtime seems to return E_FAIL and not E_INVALIDARG.

Reply:
Unspecified formats are ones that are not listed in the documentation at all. So if a driver supports them, it can return S_OK but otherwise it will return E_FAIL.

Reply:
However, there is another test that is unmodified in this PR test_check_feature_support that checks for R1_UNORM specifically for E_INVALIDARG and it passes. I am looking in to what the difference is because in this test_format_support, it returns E_FAIL.

```
        if (format == DXGI_FORMAT_R1_UNORM)
        {
            ok(hr == E_INVALIDARG, "Got unexpected hr %#x.\n", hr);
            continue;
        }
```

Reply:
R1_UNORM returns E_FAIL on AMD Windows at least. Current code passes test on Windows.

Reply:
Found the difference. In test_check_feature_support, it is querying D3D12_FEATURE_FORMAT_INFO which is correctly returning E_INVALIDARG when the format is not supported. In test_format_support, it is querying D3D12_FEATURE_FORMAT_SUPPORT which should return E_FAIL when the format is not supported or S_OK when it is.